vrclient_x64/vrclient_x64/vrclient_main.c

```diff
@@ -744,79 +769,171 @@ static EVRCompositorError ivrcompositor_submit_wined3d(
-    vkdata.m_nSampleCount = image_info.samples;
-
-    if (flags & (Submit_TextureWithPose | Submit_TextureWithDepth))
-        FIXME("Unhandled flags %#x.\n", flags & (Submit_TextureWithPose | Submit_TextureWithDepth));
-
-    vktexture = *texture;
-    vktexture.handle = &vkdata;
-    vktexture.eType = TextureType_Vulkan;
-
-    subresources.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
-    subresources.baseMipLevel = 0;
-    subresources.levelCount = image_info.mipLevels;
-    subresources.baseArrayLayer = 0;
-    subresources.layerCount = image_info.arrayLayers;
-
-    dxvk_device->lpVtbl->TransitionSurfaceLayout(dxvk_device, dxvk_surface, &subresources,
-            image_layout, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
-    dxvk_device->lpVtbl->FlushRenderingCommands(dxvk_device);
-    dxvk_device->lpVtbl->LockSubmissionQueue(dxvk_device);
-
-    err = cpp_func(linux_side, eye, &vktexture, bounds, flags);
-
-    dxvk_device->lpVtbl->ReleaseSubmissionQueue(dxvk_device);
-    dxvk_device->lpVtbl->TransitionSurfaceLayout(dxvk_device, dxvk_surface, &subresources,
-            VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, image_layout);
-
-    dxvk_device->lpVtbl->Release(dxvk_device);
-    dxvk_surface->lpVtbl->Release(dxvk_surface);
-    return err;
+static EVRCompositorError ivrcompositor_submit_dxvk(
+        EVRCompositorError (*cpp_func)(void *, EVREye, Texture_t *, VRTextureBounds_t *, EVRSubmitFlags),
+        void *linux_side, EVREye eye, Texture_t *texture, VRTextureBounds_t *bounds, EVRSubmitFlags flags,
+        unsigned int version, struct compositor_data *user_data, IDXGIVkInteropSurface *dxvk_surface)
+{
+	struct VRVulkanTextureData_t vkdata;
+	IDXGIVkInteropDevice *dxvk_device;
+	struct Texture_t vktexture;
+
+	VkImageLayout image_layout;
+	VkImageCreateInfo image_info;
+	VkImageSubresourceRange subresources;
+
+	EVRCompositorError err;
+
+	vktexture = vrclient_translate_texture_dxvk(texture, &vkdata, dxvk_surface, &dxvk_device, &image_layout, &image_info);
+	
+	subresources = vrclient_dxvk_vkimagecreateinfo_to_subresources(&image_info);
+	if (flags & (Submit_TextureWithPose | Submit_TextureWithDepth))
+		FIXME("Unhandled flags %#x.\n", flags & (Submit_TextureWithPose | Submit_TextureWithDepth));
+
+	dxvk_device->lpVtbl->TransitionSurfaceLayout(dxvk_device, dxvk_surface, &subresources,
+	        image_layout, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
+	dxvk_device->lpVtbl->FlushRenderingCommands(dxvk_device);
+	dxvk_device->lpVtbl->LockSubmissionQueue(dxvk_device);
+	
+	err = cpp_func(linux_side, eye, &vktexture, bounds, flags);
+	
+	dxvk_device->lpVtbl->ReleaseSubmissionQueue(dxvk_device);
+	dxvk_device->lpVtbl->TransitionSurfaceLayout(dxvk_device, dxvk_surface, &subresources,
+	        VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, image_layout);
+	
+	dxvk_device->lpVtbl->Release(dxvk_device);
+	dxvk_surface->lpVtbl->Release(dxvk_surface);
+	return err;
 }
 #endif
 
+static EVROverlayError ivroverlay_set_overlay_texture_vulkan(
```

Comment:
Note: this seems simple, but I was unable to test as I don't know of a Vulkan VR program for windows that actually uses the `VROverlay` API from OpenVR.